### PR TITLE
Changes in class libraries to enable Windows Runtime event support using Mono's mscorlib

### PR DESCRIPTION
**Pull request by TautvydasZilys**
This PR makes minimal changes needed to make Windows Runtime events work when using Mono's mscorlib. It mostly replaces not implemented classes and methods with reference source, filling in the gaps with glue where needed (extra methods on System.Runtime.CompilerServices.ConditionalWeakTable, new icalls in System.Runtime.InteropServices.WindowsRuntime.UnsafeNativeMethods and System.Runtime.InteropServices.Marshal).

I wasn't quite sure what to do with the new icalls - I didn't add them to the native runtime side so they now throw MissingMethodException if called in Mono runtime. However, I don't think it's possible to reach these icalls in Mono runtime at the moment, as it would first require referencing winmd files that expose events and calling add/remove methods on that, which (AFAIK) is not supported in Mono runtime today.

### CODE REVIEW

**Comment by marek-safar** [comment]
approve

File: mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs [hunk 1]
```diff
@@ -1722,5 +1722,31 @@ public static IntPtr GetFunctionPointerForDelegate (Delegate d)
 		internal static void SetLastWin32Error (int error)
 		{
 		}
+
+		// Copied from referencesource/mscorlib/system/runtime/interopservices/marshal.cs
```
Comment (TautvydasZilys) [inline mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs]: Will fix.

Comment (TautvydasZilys) [inline mcs/class/corlib/System.Runtime.InteropServices/Marshal.cs]: Fixed

**Comment by marek-safar** [comment]
Could you address the build failures

**Comment by TautvydasZilys** [comment]
Working on it.

File: mcs/class/referencesource/mscorlib/system/type.cs [hunk 2]
```diff
@@ -1249,7 +1249,6 @@ public virtual Type[] GetGenericParameterConstraints()
             get {return IsCOMObjectImpl();}
         }
 
-#if FEATURE_COMINTEROP || MONO_COM
```
Comment (marek-safar) [inline mcs/class/referencesource/mscorlib/system/type.cs]: Please don't remove referencesource defines

Comment (TautvydasZilys) [inline mcs/class/referencesource/mscorlib/system/type.cs]: What do you suggest I do then? The other reference source code is calling into this method, and it does not compile when MONO_COM is not defined.

Comment (marek-safar) [inline mcs/class/referencesource/mscorlib/system/type.cs]: Where are we doing that?

Comment (marek-safar) [inline mcs/class/referencesource/mscorlib/system/type.cs]: What's the problem with that line? All methods it calls are always available.

Comment (TautvydasZilys) [inline mcs/class/referencesource/mscorlib/system/type.cs]: "if (type.IsWindowsRuntimeObject && type.IsImport)"

IsWindowsRuntimeObject is defined behind a define removed.

Comment (marek-safar) [inline mcs/class/referencesource/mscorlib/system/type.cs]: Right, I missed that one. Better to fix `windowsruntimemarshal.cs` to match it

Comment (TautvydasZilys) [inline mcs/class/referencesource/mscorlib/system/type.cs]: Okay.

**Pull request merged**